### ISSUE TITLE
Cross-scene hierarchy is unchecked, and unloadScene leaves dangling links

### ISSUE THREAD

**Issue by sunarya-thito** [sunarya-thito/good#102]
Nothing in `hierarchy.dart` knows scenes exist — grep the file for scene or
slot and there are **zero hits**. All four operations permit cross-scene edges,
and `unloadScene` never unlinks them, so unloading one scene arms a crash in
another.

## Every operation permits it

`addChild` (`hierarchy.dart:194`), `adopt` (`:219`), `removeChild` (`:239`),
`unlinkChild` (`:261`) check archetype membership and cycles, never scene
membership. `SceneStruct.addEntityIn` (`scene.dart:336-362`) checks
`identical(prefab.scene, this)` for the *prefab* and never compares the `parent`
argument's scene to the spawning slot.

Probed, all four:

| | |
|---|---|
| `addChild` across scenes | succeeds, links written both ways |
| `adopt` from scene B into scene A | succeeds; `child.sceneSlot=1 pa.sceneSlot=0` |
| `removeChild` across scenes | succeeds — **destroys a scene-B row from a scene-A parent** |

## The damage lands at unload

`unmountEntitiesOf` (`scene.dart:442-464`) only *fires events* for rows on the
doomed slot. It does not destroy or detach them, and nothing walks surviving
scenes looking for links pointing into the freed pages.

**Unload the child's scene** — the parent keeps naming the freed row:

```
after unloading child scene, parent.firstChild = 844424930131968 (still set: true)
parent.destroy()  ->  Bad state: Entity 844424930131968 names page 0 of
                      archetype 3 (Pet), and that page has been freed
```

`destroy()` throws **part way through its subtree walk**, having already fired
half its unmount events. A listener that allocates per entity — a Box2D body, a
native handle — is left holding the ones it was never told about.

**Unload the parent's scene** — the child survives holding `parent = <freed
row>`, and both `child<Child>().detach()` and `child.destroy()` then throw,
because both route through `unlinkChild`, which writes into the freed page. The
child cannot be cleaned up by any public means.

So an ordinary "unload the pause menu" detonates on the level scene's next
`destroy()`.

## One piece of good news, and it should not be traded away

Page indices are never recycled after unload (`archetype.dart:428-435` nulls the
slot permanently), so a stale cross-scene handle throws a clear `StateError`
rather than silently aliasing the next scene's row. Probed:

```
fresh pet = 844429225099264, stale link = 844424930131968, equal: false
```

That is why this bug is loud rather than silent. `archetype.dart:620-630`
already records what it costs — a hard cap of 65 536 pages per archetype across
load/unload churn — and names the fix. **Whoever addresses that cap must keep
this property**, or this issue becomes silent corruption instead of a throw.

## Fix, in two parts

- **`assert(child.sceneSlot == entity.sceneSlot)`** on the four edit
  operations. Debug-only, per the release-check rule — these are programmer
  errors that cannot vary at runtime.
- **A detach pass in `unmountEntitiesOf`** that unlinks inbound edges from
  surviving scenes before the pages are freed. This half cannot be an assert;
  it is the actual repair.

Consider whether a cross-scene parent should be permitted at all. A subtree that
spans scenes has no coherent answer for "which scene owns this entity", and
`Entity.scene` (`scene.dart:758-771`) answers per row, so a parent and child can
legitimately disagree. Forbidding it outright is simpler than making unload
correct for it.

**Done when** cross-scene parenting fails in debug, and unloading a scene cannot
leave a surviving scene holding a link into freed pages.

**Comment by sunarya-thito** [comment]
Fixed in e4b5e99.

Both halves are in:

- `ParentAccessor._sameScene` rejects an edge whose ends are in different scene slots, asserted on all four edit operations (`addChild`, `adopt`, `removeChild`, `unlinkChild`), and `SceneStruct._attachable` now takes the spawning slot so the spawn path refuses a foreign parent before it allocates the row. All of it sits inside existing asserts and compiles out of release.
- `unmountEntitiesOf` calls `_detachLinksLeavingScene`, which walks the doomed slot in both directions - a doomed child with a surviving parent, and a doomed parent with surviving children - and splices each crossing edge before the pages are freed.

The surviving side loses the link and nothing else: a surviving parent goes childless, a surviving child becomes an unparented root still holding its own subtree. Destroying the surviving side would fire unmount events for entities of a scene that is not unloading, from inside the unload of one that is. The pass runs after all unmount events rather than interleaved with them, so a listener reads the hierarchy as it stood.

Page indices are still never recycled after unload - `archetype.dart` is untouched, so a stale cross-scene handle still throws rather than aliasing.

`packages/good/test/cross_scene_hierarchy_test.dart` adds 9 tests: 548 -> 557, with `goo2d` (222) and `goo2d_physics_box2d` (64) unchanged. Each test was proven able to fail by eight separate mutations of the fix, including disabling each unload direction on its own, so neither covers for the other.
